File: app/src/main/java/com/popkter/roundedimageview/RoundedImageView.java
package com.popkter.roundedimageview;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.BitmapShader;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Shader;
import android.graphics.drawable.BitmapDrawable;
import android.graphics.drawable.Drawable;
import android.util.AttributeSet;
import com.popkter.colorpickerview.R;

import androidx.appcompat.widget.AppCompatImageView;


/**
 * 引用自 https://blog.csdn.net/explorerqp/article/details/95336023
 */
public class RoundedImageView extends AppCompatImageView {

    private static final int DEFAULT_BORDER_WIDTH = 0;
    private static final int DEFAULT_BORDER_COLOR = Color.TRANSPARENT;


    private final int mBorderColor;
    private final int mBorderWidth;

    private final Paint paintImage;
    private final Paint paintBorder;
    /**
     * 圆角的幅度
     **/
    private final float mRadius;
    /**
     * 是否是圆形
     **/
    private final boolean mIsCircle;

    public RoundedImageView(final Context context) {
        this(context, null);
    }

    public RoundedImageView(Context context, AttributeSet attrs) {
        this(context, attrs, 0);
    }

    public RoundedImageView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);

        this.setScaleType(ScaleType.FIT_XY);
        // 获取自定属性配置
        TypedArray ta = context.obtainStyledAttributes(attrs,
                R.styleable.rounded_image_view, defStyle, 0);
        mRadius = ta.getDimensionPixelSize(R.styleable.rounded_image_view_radius, 0);
        mIsCircle = ta.getBoolean(R.styleable.rounded_image_view_circle, false);
        mBorderColor = ta.getColor(R.styleable.rounded_image_view_border_color, DEFAULT_BORDER_COLOR);
        mBorderWidth = ta.getDimensionPixelSize(R.styleable.rounded_image_view_border_width, DEFAULT_BORDER_WIDTH);
        ta.recycle();
        paintImage = new Paint();
        paintImage.setAntiAlias(true);
        paintBorder = new Paint();
        paintBorder.setAntiAlias(true);
    }


    @Override
    public void onDraw(Canvas canvas) {
        // 去点padding值
        int viewWidth = canvas.getWidth() - getPaddingLeft() - getPaddingRight();
        int viewHeight = canvas.getHeight() - getPaddingTop() - getPaddingBottom();
        // 获取iamgeView 中的drawable 并转成bitmap
        Bitmap image = drawableToBitmap(getDrawable());
        // 对获取到的bitmap 按照当前imageView的宽高进行缩放
        Bitmap reSizeImage = reSizeImage(image, viewWidth, viewHeight);
        int imgWidth = reSizeImage.getWidth();
        int imgHight = reSizeImage.getHeight();

        paintBorder.setColor(mBorderColor);
        paintBorder.setStrokeWidth(mBorderWidth);
        paintBorder.setStyle(Paint.Style.FILL_AND_STROKE);

        // 判断当前需要绘制圆角还是圆
        if (mIsCircle) {
            //画边线
            canvas.drawCircle(viewWidth / 2, viewHeight / 2, (Math.min(viewWidth, viewHeight) / 2) - mBorderWidth / 2, paintBorder);
            //设置画笔
            BitmapShader bitmapShader = new BitmapShader(reSizeImage, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP);
            paintImage.setShader(bitmapShader);
            //画布移动(x,y),移动量为边框宽度
            canvas.translate(mBorderWidth, mBorderWidth);
            canvas.drawCircle(imgWidth / 2, imgHight / 2, Math.min(imgWidth, imgHight) / 2, paintImage);

        } else {
            //画边线
            RectF rectB = new RectF(mBorderWidth / 2, mBorderWidth / 2, viewWidth - mBorderWidth / 2, viewHeight - mBorderWidth / 2);
            canvas.drawRoundRect(rectB, mRadius, mRadius, paintBorder);

            BitmapShader bitmapShader = new BitmapShader(reSizeImage, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP);
            paintImage.setShader(bitmapShader);

            RectF rect = new RectF(0, 0, imgWidth, imgHight);
            //弧度半径(有边框无圆角)
            float radius = mBorderWidth == 0 ? mRadius : 0;
            //画布移动(x,y),移动量为边框宽度
            canvas.translate(mBorderWidth, mBorderWidth);
            canvas.drawRoundRect(rect, radius, radius, paintImage);
        }
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        int width = MeasureSpec.getSize(widthMeasureSpec);
        int height = MeasureSpec.getSize(heightMeasureSpec);
        setMeasuredDimension(width, height);
    }

    /**
     * drawable转bitmap
     *
     * @param drawable
     * @return
     */
    private Bitmap drawableToBitmap(Drawable drawable) {
        if (drawable == null) {
            return null;
        } else if (drawable instanceof BitmapDrawable) {
            return ((BitmapDrawable) drawable).getBitmap();
        }

        Bitmap bitmap = Bitmap.createBitmap(drawable.getIntrinsicHeight(),
                drawable.getIntrinsicHeight(), Bitmap.Config.ARGB_8888);
        Canvas canvas = new Canvas(bitmap);
        drawable.setBounds(0, 0, canvas.getWidth(), canvas.getHeight());
        drawable.draw(canvas);
        return bitmap;
    }

    /**
     * 重设Bitmap的宽高
     *
     * @param bitmap
     * @param newWidth
     * @param newHeight
     * @return
     */
    private Bitmap reSizeImage(Bitmap bitmap, int newWidth, int newHeight) {
        int width = bitmap.getWidth();
        int height = bitmap.getHeight();

        //取得正方形边长
        int sideL = Math.min(width, height);
        //取得原点轴坐标
        int x = 0;
        int y = 0;
        if (width > height) {
            x = (width - height) / 2;
        } else {
            y = (height - width) / 2;
        }

        Bitmap bitmapSquare = Bitmap.createBitmap(bitmap, x, y, sideL, sideL);

        width = bitmapSquare.getWidth();
        height = bitmapSquare.getHeight();

        // 计算出缩放比
        float scaleHeight = (float) (newHeight - (mBorderWidth * 2)) / height;
        float scaleWidth = (float) (newWidth - (mBorderWidth * 2)) / width;
        // 矩阵缩放bitmap
        Matrix matrix = new Matrix();
        matrix.postScale(scaleWidth, scaleHeight);
        return Bitmap.createBitmap(bitmapSquare, 0, 0, width, height, matrix, true);
    }
}

File: app/src/main/java/com/popkter/roundedimageview/RoundedImageView.kt
package com.popkter.roundedimageview

import android.content.Context
import android.graphics.*
import android.graphics.drawable.BitmapDrawable
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import androidx.appcompat.widget.AppCompatImageView
import com.popkter.colorpickerview.R


class RoundedImageView @JvmOverloads constructor(
    context: Context,
    attrs: AttributeSet? = null,
    defStyle: Int = 0
) :
    AppCompatImageView(context, attrs, defStyle) {

    companion object {
        private const val DEFAULT_BORDER_WIDTH = 0
        private const val DEFAULT_BORDER_COLOR = Color.WHITE
    }

    private val mBorderColor: Int
    private val mBorderWidth: Int
    private val paintImage: Paint
    private val paintBorder: Paint

    /**
     * 圆角的幅度
     */
    private val mRadius: Float

    /**
     * 是否是圆形
     */
    private val mIsCircle: Boolean

    init {
        this.scaleType = ScaleType.FIT_XY
        // 获取自定属性配置
        val ta = context.obtainStyledAttributes(attrs, R.styleable.rounded_image_view, defStyle, 0)
        mRadius = ta.getDimensionPixelSize(R.styleable.rounded_image_view_radius, 0).toFloat()
        mIsCircle = ta.getBoolean(R.styleable.rounded_image_view_circle, false)
        mBorderColor = ta.getColor(R.styleable.rounded_image_view_border_color, DEFAULT_BORDER_COLOR)
        mBorderWidth = ta.getDimensionPixelSize(R.styleable.rounded_image_view_border_width, DEFAULT_BORDER_WIDTH)
        ta.recycle()
        paintImage = Paint()
        paintImage.isAntiAlias = true
        paintBorder = Paint()
        paintBorder.isAntiAlias = true
    }

    public override fun onDraw(canvas: Canvas) {
        // 去点padding值
        val viewWidth = width - paddingLeft - paddingRight
        val viewHeight = height - paddingTop - paddingBottom
        // 获取ImageView 中的drawable 并转成bitmap
        val image = drawableToBitmap(drawable)
        // 对获取到的bitmap 按照当前imageView的宽高进行缩放
        val reSizeImage = reSizeImage(image, viewWidth, viewHeight)
        val imgWidth = reSizeImage.width
        val imgHeight = reSizeImage.height
        paintBorder.color = mBorderColor
        paintBorder.strokeWidth = mBorderWidth.toFloat()
        paintBorder.style = Paint.Style.FILL_AND_STROKE

        // 判断当前需要绘制圆角还是圆
        if (mIsCircle) {
            //画边线
            canvas.drawCircle(
                (viewWidth / 2).toFloat(),
                (viewHeight / 2).toFloat(),
                (viewWidth.coerceAtMost(viewHeight) / 2 - mBorderWidth / 2).toFloat(),
                paintBorder
            )
            //设置画笔
            val bitmapShader = BitmapShader(reSizeImage, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
            paintImage.shader = bitmapShader
            //画布移动(x,y),移动量为边框宽度
            canvas.translate(mBorderWidth.toFloat(), mBorderWidth.toFloat())
            canvas.drawCircle(
                (imgWidth / 2).toFloat(),
                (imgHeight / 2).toFloat(),
                (imgWidth.coerceAtMost(imgHeight) / 2).toFloat(),
                paintImage
            )
        } else {
            //画边线
            val rectB = RectF(
                (mBorderWidth / 2).toFloat(),
                (mBorderWidth / 2).toFloat(),
                (viewWidth - mBorderWidth / 2).toFloat(),
                (viewHeight - mBorderWidth / 2).toFloat()
            )
            canvas.drawRoundRect(rectB, mRadius, mRadius, paintBorder)
            val bitmapShader = BitmapShader(reSizeImage, Shader.TileMode.CLAMP, Shader.TileMode.CLAMP)
            paintImage.shader = bitmapShader
            val rect = RectF(0f, 0f, imgWidth.toFloat(), imgHeight.toFloat())
            //弧度半径(有边框无圆角)
            val radius: Float = if (mBorderWidth == 0) mRadius else 0F
            //画布移动(x,y),移动量为边框宽度
            canvas.translate(mBorderWidth.toFloat(), mBorderWidth.toFloat())
            canvas.drawRoundRect(rect, radius, radius, paintImage)
        }
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        val width = MeasureSpec.getSize(widthMeasureSpec)
        val height = MeasureSpec.getSize(heightMeasureSpec)
        setMeasuredDimension(width, height)
    }

    /**
     * drawable转bitmap
     *
     * @param drawable
     * @return
     */
    private fun drawableToBitmap(drawable: Drawable?): Bitmap? {
        if (drawable == null) {
            return null
        } else if (drawable is BitmapDrawable) {
            return drawable.bitmap
        }
        val bitmap = Bitmap.createBitmap(
            drawable.intrinsicHeight,
            drawable.intrinsicHeight, Bitmap.Config.ARGB_8888
        )
        val canvas = Canvas(bitmap)
        drawable.setBounds(0, 0, canvas.width, canvas.height)
        drawable.draw(canvas)
        return bitmap
    }

    /**
     * 重设Bitmap的宽高
     *
     * @param bitmap
     * @param newWidth
     * @param newHeight
     * @return
     */
    private fun reSizeImage(bitmap: Bitmap?, newWidth: Int, newHeight: Int): Bitmap {
        var width = bitmap!!.width
        var height = bitmap.height

        //取得正方形边长
        val sideL = width.coerceAtMost(height)
        //取得原点轴坐标
        var x = 0
        var y = 0
        if (width > height) {
            x = (width - height) / 2
        } else {
            y = (height - width) / 2
        }
        val bitmapSquare = Bitmap.createBitmap(bitmap, x, y, sideL, sideL)
        width = bitmapSquare.width
        height = bitmapSquare.height

        // 计算出缩放比
        val scaleHeight = (newHeight - mBorderWidth * 2).toFloat() / height
        val scaleWidth = (newWidth - mBorderWidth * 2).toFloat() / width
        // 矩阵缩放bitmap
        val matrix = Matrix()
        matrix.postScale(scaleWidth, scaleHeight)
        return Bitmap.createBitmap(bitmapSquare, 0, 0, width, height, matrix, true)
    }
}